[NEMO-66] Move /tests/scheduler into /runtime/master/scheduler/test

johnyangk opened the pull request:
JIRA: [NEMO-66: Move /tests/scheduler into /runtime/master/scheduler/test](https://issues.apache.org/jira/projects/NEMO/issues/NEMO-66)

**Major changes:**
- Moves /tests/scheduler into /runtime/master/scheduler/test
- Creates /runtime/testplans that depend on /compiler/optimizer
- Makes /runtime/master depend on /runtime/testplans

**Minor changes to note:**
- Refactors scheduler unit tests to share code for generating physical plans
- Refactors RuntimeTestUtil
- Refactors TestPolicy 

**Tests for the changes:**
- This PR only moves tests from one directory to another, and the tests continue to pass.

**Other comments:**
- N/A

resolves [NEMO-66](https://issues.apache.org/jira/projects/NEMO/issues/NEMO-66)

johnyangk commented on pom.xml:
```diff
@@ -63,6 +63,7 @@ limitations under the License.
         <module>runtime/executor</module>
         <module>runtime/master</module>
         <module>runtime/driver</module>
+        <module>runtime/testplans</module>
```
Are you suggesting the following?

runtime/runtimetest/testpolicy
runtime/runtimetest/testplan

Can you also give me more guide about refactoring the current package into two packages? Which classes/methods do you think belong to which package?

johnyangk replied:
@jeongyooneo ping :)

jeongyooneo replied:
1. Splitting into `testpolicy` and `testplan`
> Are you suggesting the following?
runtime/runtimetest/testpolicy
runtime/runtimetest/testplan

Yes, right.

2. Splitting classes
Putting `BasicPullPoilcy` and `BasicPushPolicy` into `testpolicy` and `TestPlanGenerator` to `testplan`?

johnyangk commented:
@jeongyooneo I've resolved the conflicts with the master branch. Pleas take a look. :)

johnyangk commented:
@sanha Can you review and merge? :)

johnyangk commented:
@jeongyooneo Thanks for the comment. While separating the `testplans` package into two as per your suggestion, I've come to realize that the `BasicPushPolicy` and the `BasicPullPolicy` really should go into the `optimizer` package. So I moved those classes to the `optimizer` package.

I've also renamed `testplans` to `plangenerator`, and refactored `TestPlanGenerator` to make its public methods only deal with physical plans, and hide IRs and policies from the tests in the other `runtime` package. 

At present `TestPlanGenerator` can generate three specific types of plans used in the current scheduler tests, but we can later extend this class to generate many random, complex DAGs to be used by not only the scheduler test but also the test for the executors.

johnyangk commented on runtime/common/src/main/java/edu/snu/nemo/runtime/common/optimizer/pass/runtime/DataSkewRuntimePass.java:
```diff
@@ -106,7 +106,8 @@ public PhysicalPlan apply(final PhysicalPlan originalPlan, final Map<String, Lis
             .map(pair -> pair.left())
             .max(Integer::compareTo)
-            .orElseThrow(() -> new DynamicOptimizationException("Cannot find max hash value in a block.")))
+            .<DynamicOptimizationException>orElseThrow(
```
I've done explicit typing here to avoid the JDK bug reported in
https://stackoverflow.com/a/40865318
